{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ca22c29c-e6ae-418b-8292-95be9e4d20b4",
   "metadata": {},
   "source": [
    "# Bioinformatics Homework Problems\n",
    "To get credit for this assignment, you must:\n",
    " * (1) complete complete each of codes listed below,\n",
    " * (2) update your README to include the requested figures,\n",
    " * (3) push your resulting repository to the GitHub.\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "81b29390-e55b-4af1-94a3-15225881754c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\sherl\\anaconda3\\envs\\BIOM421\\Lib\\site-packages\\Bio\\pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Preliminary Codes -- DO NOT CHANGE ANYTHING IN THIS SECTION\n",
    "import pickle, os, sys, threading, Bio\n",
    "import numpy as np\n",
    "from Bio import Entrez, SeqIO, pairwise2, AlignIO, Phylo\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqUtils import ProtParam\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
    "\n",
    "q1 = 'xxx'\n",
    "q2 = 'XXX'\n",
    "q3 = 'XXX'\n",
    "q4 = 'XXX'\n",
    "q5 = 'XXX'\n",
    "q6 = 'XXX'\n",
    "q7 = 'XXX'\n",
    "q8 = 'XXX'\n",
    "\n",
    "accessionNumbers = ['NC_026437','AF033819','NC_001474','NC_002549',\n",
    "                   'NC_001526','NC_003977','NC_001348','NC_001803',\n",
    "                   'NC_012532','NC_001498','NC_001806','NC_001436','NC_001959']\n",
    "\n",
    "j = int(np.random.random()*13)\n",
    "accession = accessionNumbers[j]\n",
    "fnDat = 'Hwk4Data.pkl'\n",
    "dataDict = {'accession':accession}\n",
    "with open(fnDat, 'wb') as file:\n",
    "    pickle.dump(dataDict, file)\n",
    "\n",
    "def saveResults():\n",
    "    fn = 'Hwk4Answers.pkl'\n",
    "    answerDict = {'q1':q1,'q2':q2,'q3':q3,'q4':q4,'q5':q5,'q6':q6,'q7':q7,'q8':q8}\n",
    "    with open(fn, 'wb') as file:\n",
    "        pickle.dump(answerDict, file)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6bd9706c-7afb-4e44-aa26-99423920eeff",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The accession number for your sequence is NC_001806\n"
     ]
    }
   ],
   "source": [
    "#Part 1 (Basic Sequence Manupulation) -- For the accession number \"accession\", find the sequence information, and fill in the following blanks:\n",
    "\n",
    "print(f'The accession number for your sequence is {accession}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1d29f8d1-c30e-47af-a649-a874c8e5776c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n",
      "q1 = 152222, \n",
      "q2 = AGCCCGGGCCCCCCGCGGGCGCGCGCGCGCGCAAAAAAGGCGGGCGGCGGTCCGGGCGGCGTGCGCGCGCGCGGCGGGCGTGGGGGGCGGGGCCGCGGGA, \n",
      "q3 = {Seq('AG'): 7210, Seq('GC'): 16476, Seq('CC'): 18566, Seq('CG'): 17918, Seq('GG'): 19353, Seq('CA'): 8183, Seq('AA'): 4878, Seq('GT'): 8538, Seq('TC'): 8229, Seq('TG'): 8049, Seq('GA'): 8163, Seq('AC'): 8168, Seq('CT'): 6771, Seq('TT'): 4750, Seq('TA'): 3000, Seq('AT'): 3969}, \n",
      "q4 = {Seq('AGC'): 2276, Seq('GCC'): 5871, Seq('CCC'): 7003, Seq('CCG'): 5808, Seq('CGG'): 5985, Seq('GGG'): 7429, Seq('GGC'): 5985, Seq('CGC'): 6059, Seq('GCG'): 6236, Seq('GCA'): 2201, Seq('CAA'): 1374, Seq('AAA'): 1372, Seq('AAG'): 1273, Seq('AGG'): 2753, Seq('GGT'): 2959, Seq('GTC'): 2986, Seq('TCC'): 2998, Seq('CGT'): 3054, Seq('GTG'): 2750, Seq('TGC'): 2156, Seq('TGG'): 3186, Seq('GGA'): 2980, Seq('GAG'): 2566, Seq('CCA'): 3283, Seq('CAG'): 2784, Seq('AGA'): 1307, Seq('GAC'): 2799, Seq('ACC'): 2694, Seq('CGA'): 2820, Seq('GAA'): 1459, Seq('AAC'): 1632, Seq('ACG'): 3012, Seq('ACA'): 1616, Seq('CAC'): 2690, Seq('TCG'): 2862, Seq('CCT'): 2472, Seq('CTT'): 1138, Seq('TTT'): 1320, Seq('TTA'): 644, Seq('TAA'): 673, Seq('ACT'): 846, Seq('CTA'): 552, Seq('TAG'): 586, Seq('AGT'): 874, Seq('GTT'): 1688, Seq('TAC'): 1047, Seq('CTC'): 2561, Seq('TCT'): 1286, Seq('CTG'): 2520, Seq('CAT'): 1335, Seq('ATG'): 1368, Seq('GCT'): 2167, Seq('TTG'): 1411, Seq('TGT'): 1651, Seq('TCA'): 1083, Seq('ATC'): 1307, Seq('GTA'): 1114, Seq('TTC'): 1375, Seq('GAT'): 1339, Seq('TGA'): 1056, Seq('ATA'): 690, Seq('TAT'): 694, Seq('AAT'): 601, Seq('ATT'): 604}, \n",
      "q5 = ATGTCTCCCGAGCGTCAAAATCCGGGGGGGGGCGGCGACGGTCAAGGGGAGGGTTGGAGACCGGGGTTGGGGAATGAATCCCTCCCCTTCACCGACAACCCCCCGGGTAACCACGGGGTGCCGATGAACCCCGGCGGCCGGCAACGCGGGGTCCCTGCGAGAGGCACAGATGCTTACGGTCAGGTGCTCCGGGTCGGGTGCGTCTGGTATGCGGTTGGTATATGTACACTTTACCTGGGGGCGTGCCGGTCCGCCCCAGCCCCTCCCACGCCCCGCGCGTCATCAGCCGGTGGGCGTGGCCGCTATTATAAAAAAAGTGAGAACGCGAAGCGTTCGCACTTTGTCCTAATAATATATATATTATTAGGACAAAGTGCGAACGCTTCGCGTTCTCACTTTTTTTATAATAGCGGCCACGCCCACCGGCTACGTCACTCTCCTGTCGGCCGCCGGCGGTCCATAAGCCCGGCCGGCCGGGCCGACGCGAATAAACCGGGCCGCCGGCCGGGGCGCCGCGCAGCAGCTCGCCGCCCGGATCCGCCAGACAAACAAGGCCCTTGCACATGCCGGCCCGGGCGAGCCTGGGGGTCCGGTAATTTTGCCATCCCACCCAAGCGGCTTTTTGGGTTTTTCTCTTCCCCCCTCCCCACATTCCCCTCTTTAGGGGTTCGGGTGGGAACAACCGCGATGTTTTCCGGTGGCGGCGGCCCGCTGTCCCCCGGAGGAAAGTCGGCGGCCAGGGCGGCGTCCGGGTTTTTTGCGCCCGCCGGCCCTCGCGGAGCCAGCCGGGGACCCCCGCCTTGTTTGAGGCAAAACTTTTACAACCCCTACCTCGCCCCAGTCGGGACGCAACAGAAGCCGACCGGGCCAACCCAGCGCCATACGTACTATAGCGAATGCGATGAATTTCGATTCATCGCCCCGCGGGTGCTGGACGAGGATGCCCCCCCGGAGAAGCGCGCCGGGGTGCACGACGGTCACCTCAAGCGCGCCCCCAAGGTGTACTGCGGGGGGGACGAGCGCGACGTCCTCCGCGTCGGGTCGGGCGGCTTCTGGCCGCGGCGCTCGCGCCTGTGGGGCGGCGTGGACCACGCCCCGGCGGGGTTCAACCCCACCGTCACCGTCTTTCACGTGTACGACATCCTGGAGAACGTGGAGCACGCGTACGGCATGCGCGCGGCCCAGTTCCACGCGCGGTTTATGGACGCCATCACACCGACGGGGACCGTCATCACGCTCCTGGGCCTGACTCCGGAAGGCCACCGGGTGGCCGTTCACGTTTACGGCACGCGGCAGTACTTTTACATGAACAAGGAGGAGGTCGACAGGCACCTACAATGCCGCGCCCCACGAGATCTCTGCGAGCGCATGGCCGCGGCCCTGCGCGAGTCCCCGGGCGCGTCGTTCCGCGGCATCTCCGCGGACCACTTCGAGGCGGAGGTGGTGGAGCGCACCGACGTGTACTACTACGAGACGCGCCCCGCTCTGTTTTACCGCGTCTACGTCCGAAGCGGGCGCGTGCTGTCGTACCTGTGCGACAACTTCTGCCCGGCCATCAAGAAGTACGAGGGTGGGGTCGACGCCACCACCCGGTTCATCCTGGACAACCCCGGGTTCGTCACCTTCGGCTGGTACCGTCTCAAACCGGGCCGGAACAACACGCTAGCCCAGCCGCGGGCCCCGATGGCCTTCGGGACATCCAGCGACGTCGAGTTTAACTGTACGGCGGACAACCTGGCCATCGAGGGGGGCATGAGCGACCTACCGGCATACAAGCTCATGTGCTTCGATATCGAATGCAAGGCGGGGGGGGAGGACGAGCTGGCCTTTCCGGTGGCCGGGCACCCGGAGGACCTGGTCATCCAGATATCCTGTCTGCTCTACGACCTGTCCACCACCGCCCTGGAGCACGTCCTCCTGTTTTCGCTCGGTTCCTGCGACCTCCCCGAATCCCACCTGAACGAGCTGGCGGCCAGGGGCCTGCCCACGCCCGTGGTTCTGGAATTCGACAGCGAATTCGAGATGCTGTTGGCCTTCATGACCCTTGTGAAACAGTACGGCCCCGAGTTCGTGACCGGGTACAACATCATCAACTTCGACTGGCCCTTCTTGCTGGCCAAGCTGACGGACATTTACAAGGTCCCCCTGGACGGGTACGGCCGCATGAACGGCCGGGGCGTGTTTCGCGTGTGGGACATAGGCCAGAGCCACTTCCAGAAGCGCAGCAAGATAAAGGTGAACGGCATGGTGAACATCGACATGTACGGGATTATAACCGACAAGATCAAGCTCTCGAGCTACAAGCTCAACGCCGTGGCCGAAGCCGTCCTGAAGGACAAGAAGAAGGACCTGAGCTATCGCGACATCCCCGCCTACTACGCCGCCGGGCCCGCGCAACGCGGGGTGATCGGCGAGTACTGCATACAGGATTCCCTGCTGGTGGGCCAGCTGTTTTTTAAGTTTTTGCCCCATCTGGAGCTCTCGGCCGTCGCGCGCTTGGCGGGTATTAACATCACCCGCACCATCTACGACGGCCAGCAGATCCGCGTCTTTACGTGCCTGCTGCGCCTGGCCGACCAGAAGGGCTTTATTCTGCCGGACACCCAGGGGCGATTTAGGGGCGCCGGGGGGGAGGCGCCCAAGCGTCCGGCCGCAGCCCGGGAGGACGAGGAGCGGCCAGAGGAGGAGGGGGAGGACGAGGACGAACGCGAGGAGGGCGGGGGCGAGCGGGAGCCGGAGGGCGCGCGGGAGACCGCCGGCAGGCACGTGGGGTACCAGGGGGCCAGGGTCCTTGACCCCACTTCCGGGTTTCACGTGAACCCCGTGGTGGTGTTCGACTTTGCCAGCCTGTACCCCAGCATCATCCAGGCCCACAACCTGTGCTTCAGCACGCTCTCCCTGAGGGCCGACGCAGTGGCGCACCTGGAGGCGGGCAAGGACTACCTGGAGATCGAGGTGGGGGGGCGACGGCTGTTCTTCGTCAAGGCTCACGTGCGAGAGAGCCTCCTCAGCATCCTCCTGCGGGACTGGCTCGCCATGCGAAAGCAGATCCGCTCGCGGATTCCCCAGAGCAGCCCCGAGGAGGCCGTGCTCCTGGACAAGCAGCAGGCCGCCATCAAGGTCGTGTGTAACTCGGTGTACGGGTTCACGGGAGTGCAGCACGGACTCCTGCCGTGCCTGCACGTTGCCGCGACGGTGACGACCATCGGCCGCGAGATGCTGCTCGCGACCCGCGAGTACGTCCACGCGCGCTGGGCGGCCTTCGAACAGCTCCTGGCCGATTTCCCGGAGGCGGCCGACATGCGCGCCCCCGGGCCCTATTCCATGCGCATCATCTACGGGGACACGGACTCCATCTTTGTGCTGTGCCGCGGCCTCACGGCCGCCGGGCTGACGGCCGTGGGCGACAAGATGGCGAGCCACATCTCGCGCGCGCTGTTTCTGCCCCCCATCAAACTCGAGTGCGAAAAGACGTTCACCAAGCTGCTGCTGATCGCCAAGAAAAAGTACATCGGCGTCATCTACGGGGGTAAGATGCTCATCAAGGGCGTGGATCTGGTGCGCAAAAACAACTGCGCGTTTATCAACCGCACCTCCAGGGCCCTGGTCGACCTGCTGTTTTACGACGATACCGTCTCCGGAGCGGCCGCCGCGTTAGCCGAGCGCCCCGCGGAGGAGTGGCTGGCGCGACCCCTGCCCGAGGGACTGCAGGCGTTCGGGGCCGTCCTCGTAGACGCCCATCGGCGCATCACCGACCCGGAGAGGGACATCCAGGACTTTGTCCTCACCGCCGAACTGAGCAGACACCCGCGCGCGTACACCAACAAGCGCCTGGCCCACCTGACGGTGTATTACAAGCTCATGGCCCGCCGCGCGCAGGTCCCGTCCATCAAGGACCGGATCCCGTACGTGATCGTGGCCCAGACCCGCGAGGTAGAGGAGACGGTCGCGCGGCTGGCCGCCCTCCGCGAGCTAGACGCCGCCGCCCCAGGGGACGAGCCCGCCCCCCCCGCGGCCCTGCCCTCCCCGGCCAAGCGCCCCCGGGAGACGCCGTCGCCTGCCGACCCCCCGGGAGGCGCGTCCAAGCCCCGCAAGCTGCTGGTGTCCGAGCTGGCCGAGGATCCCGCATACGCCATTGCCCACGGCGTCGCCCTGAACACGGACTATTACTTCTCCCACCTGTTGGGGGCGGCGTGCGTGACATTCAAGGCCCTGTTTGGGAATAACGCCAAGATCACCGAGAGTCTGTTAAAAAGGTTTATTCCCGAAGTGTGGCACCCCCCGGACGACGTGGCCGCGCGGCTCCGGACCGCAGGGTTCGGGGCGGTGGGTGCCGGCGCTACGGCGGAGGAAACTCGTCGAATGTTGCATAGAGCCTTTGATACTCTAGCATGA, \n",
      "q6 = MSPERQNPGGGGDGQGEGWRPGLGNESLPFTDNPPGNHGVPMNPGGRQRGVPARGTDAYGQVLRVGCVWYAVGICTLYLGACRSAPAPPTPRASSAGGRGRYYKKSENAKRSHFVLIIYILLGQSANASRSHFFYNSGHAHRLRHSPVGRRRSISPAGRADANKPGRRPGRRAAARRPDPPDKQGPCTCRPGRAWGSGNFAIPPKRLFGFFSSPLPTFPSLGVRVGTTAMFSGGGGPLSPGGKSAARAASGFFAPAGPRGASRGPPPCLRQNFYNPYLAPVGTQQKPTGPTQRHTYYSECDEFRFIAPRVLDEDAPPEKRAGVHDGHLKRAPKVYCGGDERDVLRVGSGGFWPRRSRLWGGVDHAPAGFNPTVTVFHVYDILENVEHAYGMRAAQFHARFMDAITPTGTVITLLGLTPEGHRVAVHVYGTRQYFYMNKEEVDRHLQCRAPRDLCERMAAALRESPGASFRGISADHFEAEVVERTDVYYYETRPALFYRVYVRSGRVLSYLCDNFCPAIKKYEGGVDATTRFILDNPGFVTFGWYRLKPGRNNTLAQPRAPMAFGTSSDVEFNCTADNLAIEGGMSDLPAYKLMCFDIECKAGGEDELAFPVAGHPEDLVIQISCLLYDLSTTALEHVLLFSLGSCDLPESHLNELAARGLPTPVVLEFDSEFEMLLAFMTLVKQYGPEFVTGYNIINFDWPFLLAKLTDIYKVPLDGYGRMNGRGVFRVWDIGQSHFQKRSKIKVNGMVNIDMYGIITDKIKLSSYKLNAVAEAVLKDKKKDLSYRDIPAYYAAGPAQRGVIGEYCIQDSLLVGQLFFKFLPHLELSAVARLAGINITRTIYDGQQIRVFTCLLRLADQKGFILPDTQGRFRGAGGEAPKRPAAAREDEERPEEEGEDEDEREEGGGEREPEGARETAGRHVGYQGARVLDPTSGFHVNPVVVFDFASLYPSIIQAHNLCFSTLSLRADAVAHLEAGKDYLEIEVGGRRLFFVKAHVRESLLSILLRDWLAMRKQIRSRIPQSSPEEAVLLDKQQAAIKVVCNSVYGFTGVQHGLLPCLHVAATVTTIGREMLLATREYVHARWAAFEQLLADFPEAADMRAPGPYSMRIIYGDTDSIFVLCRGLTAAGLTAVGDKMASHISRALFLPPIKLECEKTFTKLLLIAKKKYIGVIYGGKMLIKGVDLVRKNNCAFINRTSRALVDLLFYDDTVSGAAAALAERPAEEWLARPLPEGLQAFGAVLVDAHRRITDPERDIQDFVLTAELSRHPRAYTNKRLAHLTVYYKLMARRAQVPSIKDRIPYVIVAQTREVEETVARLAALRELDAAAPGDEPAPPAALPSPAKRPRETPSPADPPGGASKPRKLLVSELAEDPAYAIAHGVALNTDYYFSHLLGAACVTFKALFGNNAKITESLLKRFIPEVWHPPDDVAARLRTAGFGAVGAGATAEETRRMLHRAFDTLA*\n"
     ]
    }
   ],
   "source": [
    "# For the accession number \"accession\", find the sequence information, \n",
    "# and fill in the following blanks.  \n",
    "\n",
    "# NOTE - I recommend using multiple cells to complete this part as some of the\n",
    "# questions (especially q5) may require some work to complete.  If you do this,\n",
    "# please make sure to save your results using the saveResults() function after\n",
    "# you complete each question.\n",
    "\n",
    "q1 = 123 # the LENGTH of the sequence\n",
    "\n",
    "q2 = 'GATCGATC' # the FIRST 100 nucleotides of the sequence\n",
    "\n",
    "q3 = {'gg':123} # a DICTIONARY with the numbers of each of the 16 possible dimers\n",
    "\n",
    "q4 = {'ggg':12} # a DICTIONARY with the numbers of each of the 4**3 = 64 different trimers\n",
    "\n",
    "q5 = 'AGTC' # the largest Open Reading Frame (ORF) in the sequence (assume that you are given the CODING sequence)\n",
    "\n",
    "q6 = 'LMNOP' # the amino acid sequence corresponding to the above ORF\n",
    "\n",
    "\n",
    "# Fetching sequence information for the given accession number\n",
    "Entrez.email = 'your_email@example.com'  # Replace with your email address\n",
    "\n",
    "\n",
    "handle = Entrez.efetch(db='nucleotide', id=accession, rettype='gb', retmode='text')\n",
    "record = SeqIO.read(handle, 'genbank')\n",
    "handle.close()\n",
    "\n",
    "\n",
    "# Filling in the blanks\n",
    "\n",
    "\n",
    "# q1: the LENGTH of the sequence\n",
    "q1 = len(record.seq)\n",
    "\n",
    "\n",
    "# q2: the FIRST 100 nucleotides of the sequence\n",
    "q2 = str(record.seq[:100])\n",
    "# q3: a DICTIONARY with the numbers of each of the 16 possible dimers    \n",
    "q3 = {}\n",
    "for i in range(len(record.seq) - 1):\n",
    "   dimer = record.seq[i:i+2]\n",
    "   if dimer in q3:\n",
    "       q3[dimer] += 1\n",
    "   else:\n",
    "       q3[dimer] = 1\n",
    "\n",
    "\n",
    "# q4: a DICTIONARY with the numbers of each of the 4**3 = 64 different trimers\n",
    "q4 = {}\n",
    "for i in range(len(record.seq) - 2):\n",
    "   trimer = record.seq[i:i+3]\n",
    "   if trimer in q4:\n",
    "       q4[trimer] += 1\n",
    "   else:\n",
    "       q4[trimer] = 1\n",
    "\n",
    "\n",
    "q5 = 'AGTC' # the largest Open Reading Frame (ORF) in the sequence (assume that you are given the CODING sequence)\n",
    "\n",
    "def find_largest_orf(sequence):\n",
    "   start_codon = 'ATG'\n",
    "   stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "\n",
    "   orfs = []\n",
    "   start_positions = [i for i in range(len(sequence) - 2) if sequence[i:i+3] == start_codon]\n",
    "\n",
    "   for start_position in start_positions:\n",
    "       for i in range(start_position + 3, len(sequence) - 2, 3):\n",
    "           codon = sequence[i:i+3]\n",
    "           if codon in stop_codons:\n",
    "               orfs.append(sequence[start_position:i+3])\n",
    "               break\n",
    "\n",
    "   if not orfs:\n",
    "       return None\n",
    "\n",
    "   return str(max(orfs, key=len))  # Convert the output to a string\n",
    "\n",
    "# Finding the largest ORF in the coding sequence\n",
    "coding_sequence = record.seq \n",
    "q5 = find_largest_orf(str(coding_sequence))\n",
    "q5 = q5[:-3]  # Remove the stop codon from the end of the sequence\n",
    "\n",
    "print(type(q5))\n",
    "\n",
    "# q6: the amino acid sequence corresponding to the above ORF\n",
    "if q5:\n",
    "   q6 = str(Seq(q5).translate())\n",
    "else:\n",
    "   q6 = None\n",
    "\n",
    "\n",
    "print(f'q1 = {q1}, \\nq2 = {q2}, \\nq3 = {q3}, \\nq4 = {q4}, \\nq5 = {q5}, \\nq6 = {q6}')\n",
    "\n",
    "saveResults()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "12c3b1eb-1cbb-4f7b-ae12-632fb40f827f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence ID: seq1, Gene: DUSP1\n",
      "Sequence ID: seq2, Gene: Dusp1\n",
      "Sequence ID: seq3, Gene: DUSP1\n",
      "Sequence ID: seq4, Gene: DUSP1\n",
      "Sequence ID: seq5, Gene: DUSP1\n",
      "Sequence ID: seq6, Gene: DUSP1\n",
      "Sequence ID: seq7, Gene: DUSP1\n",
      "Sequence ID: seq8, Gene: DUSP1\n",
      "Sequence ID: seq9, Gene: DUSP1\n"
     ]
    }
   ],
   "source": [
    "#Part 2 (Alignment)\n",
    "\n",
    "# Consider the nine related mRNA sequences with the following accession numbers:\n",
    "sequences = {'seq1':'NM_004417.4',\n",
    "             'seq2':'NM_013642.3',\n",
    "             'seq3':'NM_001085359.3',\n",
    "             'seq4':'XM_038494850.1',\n",
    "             'seq5':'NM_001046452.2',\n",
    "             'seq6':'NM_001257450.2',\n",
    "             'seq7':'XM_030283971.3',\n",
    "             'seq8':'XM_002916919.4',\n",
    "             'seq9':'NM_001256075.1'\n",
    "            }\n",
    "\n",
    "q7 = 'DUSP1' # (Not AutoGraded) Determine what gene has been sequenced (all should be the same gene).  \n",
    "# You will need to find the sequence and manually read through the annotations to find the gene name.\n",
    "# Make sure to include the gene name and a brief description of it in your README file.\n",
    "# Fetch the sequences from NCBI\n",
    "def fetch_sequences(sequences):\n",
    "    Entrez.email = 'Sherly13@colostate.edu'  # Always tell NCBI who you are\n",
    "    sequence_data = {}\n",
    "    for seq_id, accession in sequences.items():\n",
    "        handle = Entrez.efetch(db='nucleotide', id=accession, rettype='gb', retmode='text')\n",
    "        record = SeqIO.read(handle, 'genbank')\n",
    "        sequence_data[seq_id] = record\n",
    "        handle.close()\n",
    "    return sequence_data\n",
    "\n",
    "\n",
    "# Determine the gene\n",
    "def determine_gene(sequence):\n",
    "    for feature in sequence.features:\n",
    "        if feature.type == 'gene':\n",
    "            return feature.qualifiers['gene'][0]\n",
    "\n",
    "\n",
    "# Fetch the sequences\n",
    "sequences = fetch_sequences(sequences)\n",
    "\n",
    "\n",
    "# Determine the gene\n",
    "for seq_id, seq in sequences.items():\n",
    "    gene = determine_gene(seq)\n",
    "    print(f'Sequence ID: {seq_id}, Gene: {gene}')\n",
    "\n",
    "\n",
    "# Q9 - (Not AutoGraded) Make a figure to visualize the alignment of these sequences,\n",
    "# and add it to the README of your homework repository. Breifly explain your results.\n",
    "\n",
    "\n",
    "# Q10 - (Not AutoGraded) Make a figure to visualize the phylogenetic variation of these sequences, \n",
    "# and add it to the README of your homework repository. Breifly explain your results.\n",
    "\n",
    "# Q11 - (Not AutoGraded) Use BLAST to search for and add at least 5 more species to your list and \n",
    "# generate a more complete phylogenetic tree. Include this in your README file and explain.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "53df801c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Organism: NM_004417.4 Homo sapiens dual specificity phosphatase 1 (DUSP1), mRNA\n",
      "Organism: NM_013642.3 Mus musculus dual specificity phosphatase 1 (Dusp1), mRNA\n",
      "Organism: NM_001085359.3 Gallus gallus dual specificity phosphatase 1 (DUSP1), mRNA\n",
      "Organism: XM_038494850.1 PREDICTED: Canis lupus familiaris dual specificity phosphatase 1 (DUSP1), mRNA\n",
      "Organism: NM_001046452.2 Bos taurus dual specificity phosphatase 1 (DUSP1), mRNA\n",
      "Organism: NM_001257450.2 Macaca mulatta dual specificity phosphatase 1 (DUSP1), mRNA\n",
      "Organism: XM_030283971.3 PREDICTED: Taeniopygia guttata dual specificity phosphatase 1 (DUSP1), mRNA\n",
      "Organism: XM_002916919.4 PREDICTED: Ailuropoda melanoleuca dual specificity phosphatase 1 (DUSP1), mRNA\n",
      "Organism: NM_001256075.1 Sus scrofa dual specificity phosphatase 1 (DUSP1), mRNA\n",
      "q7 = DUSP1, \n",
      "q8 = ['Homo sapiens', 'Mus musculus', 'Gallus gallus', 'Canis lupus familiaris', 'Bos taurus', 'Macaca mulatta', 'Taeniopygia guttata', 'Ailuropoda melanoleuca', 'Sus scrofa']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "def fetch_sequences(accession_dict):\n",
    "    Entrez.email = 'Sherly13@colostate.edu'  # Always provide your email\n",
    "    sequence_data = {}\n",
    "    for organism, accession in accession_dict.items():\n",
    "        handle = Entrez.efetch(db='nucleotide', id=accession, rettype='fasta', retmode='text')\n",
    "        record = SeqIO.read(handle, 'fasta')\n",
    "        sequence_data[organism] = {'organism': record.description, 'sequence': str(record.seq)}\n",
    "        handle.close()\n",
    "    return sequence_data\n",
    "\n",
    "# Example usage:\n",
    "sequences = {'seq1':'NM_004417.4',\n",
    "             'seq2':'NM_013642.3',\n",
    "             'seq3':'NM_001085359.3',\n",
    "             'seq4':'XM_038494850.1',\n",
    "             'seq5':'NM_001046452.2',\n",
    "             'seq6':'NM_001257450.2',\n",
    "             'seq7':'XM_030283971.3',\n",
    "             'seq8':'XM_002916919.4',\n",
    "             'seq9':'NM_001256075.1'\n",
    "            }\n",
    "\n",
    "sequence_data = fetch_sequences(sequences)\n",
    "\n",
    "for sequence_id, data in sequence_data.items():\n",
    "    print(f\"Organism: {data['organism']}\")\n",
    "    \n",
    "q8 = ['Homo sapiens','Mus musculus','Gallus gallus', 'Canis lupus familiaris', 'Bos taurus', 'Macaca mulatta', 'Taeniopygia guttata', 'Ailuropoda melanoleuca', 'Sus scrofa'] \n",
    "print(f'q7 = {q7}, \\nq8 = {q8}')\n",
    "saveResults()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a56ed55a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "from Bio import Entrez, SeqIO, pairwise2, AlignIO, Phylo\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqUtils import ProtParam\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.image as mpimg\n",
    "import webbrowser\n",
    "\n",
    "import os\n",
    "import threading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eae7180",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Align\n",
    "from Bio.Align import Alignment\n",
    "from pymsaviz import MsaViz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c96f0099",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "from pymsaviz import MsaViz\n",
    "import os\n",
    "\n",
    "def fetch_sequences(accession_dict):\n",
    "    Entrez.email = 'Sherly13@colostate.edu'  # Always provide your email\n",
    "    sequence_data = {}\n",
    "    for seq_id, accession in accession_dict.items():\n",
    "        handle = Entrez.efetch(db='nucleotide', id=accession, rettype='fasta', retmode='text')\n",
    "        record = SeqIO.read(handle, 'fasta')\n",
    "        sequence_data[seq_id] = record\n",
    "        handle.close()\n",
    "    return sequence_data\n",
    "\n",
    "# Example usage:\n",
    "sequences = {'seq1':'NM_004417.4',\n",
    "             'seq2':'NM_013642.3',\n",
    "             'seq3':'NM_001085359.3',\n",
    "             'seq4':'XM_038494850.1',\n",
    "             'seq5':'NM_001046452.2',\n",
    "             'seq6':'NM_001257450.2',\n",
    "             'seq7':'XM_030283971.3',\n",
    "             'seq8':'XM_002916919.4',\n",
    "             'seq9':'NM_001256075.1'\n",
    "            }\n",
    "\n",
    "# Fetch sequences\n",
    "sequence_data = fetch_sequences(sequences)\n",
    "\n",
    "# Create a FASTA file with fetched sequences\n",
    "seqFile = \"Sequence.fa\"\n",
    "with open(seqFile, \"w\") as handle:\n",
    "    for seq_id, record in sequence_data.items():\n",
    "        handle.write(f\">{seq_id}\\n{record.seq}\\n\")\n",
    "\n",
    "# Use a terminal command (bash) to run MUSCLE to align the sequences.\n",
    "alignedFile = \"alignment_results.fa\"\n",
    "bashCommand = f\"muscle -in {seqFile} -out {alignedFile}\"\n",
    "os.system(bashCommand)\n",
    "\n",
    "# Make and save an alignment figure\n",
    "mv = MsaViz(alignedFile, wrap_length=100, show_count=True)\n",
    "mv.savefig(\"Sequence_Alignment.png\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcd7bef2",
   "metadata": {},
   "outputs": [],
   "source": [
    "img = mpimg.imread('Sequence_Alignment.png')\n",
    "\n",
    "# Display the image\n",
    "fig = plt.figure(figsize=(200, 80), dpi=200)\n",
    "plt.imshow(img)\n",
    "plt.axis('off')  # Hide axis\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
